org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/semantictokens/SemanticTokensService.java

```diff
@@ -0,0 +1,164 @@
+		}
+		return Collections.emptyList();
+	}
+
+	/**
+	 * Semantic tokens full.
+	 *
+	 * @param document
+	 *            the document
+	 * @param resource
+	 *            the resource
+	 * @param params
+	 *            the params
+	 * @param cancelIndicator
+	 *            the cancel indicator
+	 * @return the semantic tokens
+	 */
+	public SemanticTokens semanticTokensFull(final Document document, final XtextResource resource,
+			final SemanticTokensParams params, final CancelIndicator cancelIndicator) {
+		if (document == null || resource == null) {
+			return new SemanticTokens(Collections.emptyList());
+		}
+		List<Integer> data = new ArrayList<>();
+		int lastOffset = 0;
+		int lastLine = 0;
+		try (PositionReader reader = new PositionReader(document.getContents())) {
+			List<LightweightPosition> positions = getPositions(resource, cancelIndicator);
+			positions.sort(Comparator.comparing(LightweightPosition::getOffset));
+			for (LightweightPosition lightweightPosition : positions) {
+				Integer positionTokenType = getTokenType(lightweightPosition.getIds());
+				Integer positionTokenModifiers = getTokenModifiers(lightweightPosition.getIds());
+
+				if (positionTokenType != 0 || positionTokenModifiers != 0) {
+					int deltaOffset = lightweightPosition.getOffset() - lastOffset;
+					reader.skip(deltaOffset);
+					Position position = reader.getPosition();
+					int deltaLine = position.getLine() - lastLine;
+					data.add(deltaLine); // delta line
+					lastLine = position.getLine();
+					// delta start relative to previous token if on the same line or to 0
+					data.add(deltaLine == 0 ? deltaOffset : position.getCharacter());
+					lastOffset = lightweightPosition.getOffset();
+					data.add(lightweightPosition.getLength()); // length
+					data.add(positionTokenType); // token type
+					data.add(positionTokenModifiers); // token modifiers
+				}
+			}
+		} catch (IOException e) {
+			throw new WrappedException(e);
+		}
+		return new SemanticTokens(data); // resultId not set, no delta support
+	}
+
+	public List<String> getTokenTypes() {
+		return tokenTypes;
+	}
+
+	public List<String> getTokenModifiers() {
+		return tokenModifiers;
+	}
+
+	private int getTokenType(final String[] ids) {
+		for (String id : ids) {
+			Integer index = tokenTypes.indexOf(id);
+			if (index != -1) {
+				return index + 1;
+			}
+		}
+		return 0;
+	}
+
+	private Integer getTokenModifiers(final String[] ids) {
+		int bitmask = 0;
+		for (String id : ids) {
+			Integer index = tokenModifiers.indexOf(id);
+			if (index != -1) {
+				bitmask |= 1 << index;
+			}
+		}
+		return bitmask;
+	}
+
+	private <Service> Service getService(final URI uri, final Class<Service> type) {
+		IResourceServiceProvider resourceServiceProvider = languagesRegistry.getResourceServiceProvider(uri);
+		if (resourceServiceProvider != null) {
+			return resourceServiceProvider.get(type);
+		}
+		return null;
+	}
+}
+
+/* Copyright (c) Avaloq Evolution AG */
```

Comment:
The copyright comment here does not adhere to the code style used throughout the framework. Please remove.

Reply:
Same in other files of the MR.